networkx/algorithms/coloring/greedy_coloring.py

```diff
@@ -46,29 +48,66 @@ def strategy_random_sequential(G, colors):
     return nodes
 
 
-# TODO There is a faster implementation of this algorithm: keep track of
-# the degrees and update them manually each time a node is removed.
 def strategy_smallest_last(G, colors):
     """Returns a list of the nodes of ``G``, "smallest" last.
 
-    Specifically, the node of minimum degree is repeatedly popped from
-    the graph, then this order is reversed.
+    Specifically, the degrees of each node are tracked. From this,
+    the node of minimum degree is repeatedly popped from the graph,
+    updating its neighbors' degrees, then this order is reversed. 
 
     ``G`` is a NetworkX graph. ``colors`` is ignored.
 
-    This implementation of the strategy runs in `O(n^2)` time.
+    This implementation of the strategy runs in `O(|V| + |E|)` time.
 
     This strategy is related to :func:`strategy_independent_set`: if we
     interpret each node removed as an independent set of size one, then
     this strategy chooses an independent set of size one instead of a
     maximal independent set.
 
+    A previous version of networkx noted that this faster algorithm
+    existed; this is the implementation.
+
+    ~~~~~~
+
+    Implementation details:
+    - Sets are used rather than lists for quick random-access deletes.
+    - In Python, pop() is deterministic for ints, which are the node IDs.
```

Comment:
See [here](http://stackoverflow.com/questions/10432022/in-python-is-set-pop-deterministic). Because each node's ID is hashed the same way, `set.pop()`'s selection will be consistent. If you relabel the nodes, the behavior will be different, but still consistent from Python invocation to Python invocation.